src/utils/DirectHub.cxx

```diff
@@ -0,0 +1,943 @@
+/** \copyright
+ * Copyright (c) 2020, Balazs Racz
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ *  - Redistributions of source code must retain the above copyright notice,
+ *    this list of conditions and the following disclaimer.
+ *
+ *  - Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ * \file DirectHub.cxx
+ *
+ * Optimized class for ingress and egress of write-once objects, aimed to
+ * support multi-recipient messages with fast internal fan-out but low compute
+ * overhead.
+ *
+ * @author Balazs Racz
+ * @date 9 Feb 2020
+ */
+
+// #define LOGLEVEL VERBOSE
+
+#include "utils/DirectHub.hxx"
+
+#include <algorithm>
+#include <fcntl.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <vector>
+
+#include "executor/AsyncNotifiableBlock.hxx"
+#include "executor/StateFlow.hxx"
+#include "nmranet_config.h"
+#include "utils/logging.h"
+#include "utils/socket_listener.hxx"
+
+/// This object forwards allocations to mainBufferPool. The blocks allocated
+/// here are all the same size. They are used to read bytes from a tcp socket
+/// into memory.
+DataBufferPool g_direct_hub_data_pool(
+    config_directhub_port_incoming_buffer_size());
+/// This object forwards allocations to mainBufferPool. The blocks allocated
+/// here are all the same size. They are used to render outgoing CAN packets
+/// into gridconnect format.
+DataBufferPool g_direct_hub_kbyte_pool(1024);
```

Comment:
should 1024 be moved to a constant for tuning?

Reply:
it is not needed to adjust the size of this. For adjustable large buffers we use the g_direct_hub_data_pool.